src/manifolds/StiefelEuclideanMetric.jl

```diff
@@ -0,0 +1,144 @@
+default_metric_dispatch(::Stiefel, ::EuclideanMetric) = Val(true)
+
+function decorator_transparent_dispatch(::typeof(distance), ::Stiefel, args...)
+    return Val(:intransparent)
+end
+
+@doc raw"""
+    exp(M::Stiefel, p, X)
+
+Compute the exponential map on the [`Stiefel`](@ref)`{n,k,𝔽}`() manifold `M`
+emanating from `p` in tangent direction `X`.
+
+````math
+\exp_p X = \begin{pmatrix}
+   p\\X
+ \end{pmatrix}
+ \operatorname{Exp}
+ \left(
+ \begin{pmatrix} p^{\mathrm{H}}X & - X^{\mathrm{H}}X\\
+ I_n & p^{\mathrm{H}}X\end{pmatrix}
+ \right)
+\begin{pmatrix}  \exp( -p^{\mathrm{H}}X) \\ 0_n\end{pmatrix},
+````
+
+where $\operatorname{Exp}$ denotes matrix exponential,
+$\cdot^{\mathrm{H}}$ denotes the complex conjugate transpose or Hermitian, and $I_k$ and
+$0_k$ are the identity matrix and the zero matrix of dimension $k × k$, respectively.
+"""
+exp(::Stiefel, ::Any...)
+
+function exp!(::Stiefel{n,k}, q, p, X) where {n,k}
+    return copyto!(
+        q,
+        [p X] *
+        exp([p'X -X'*X; one(zeros(eltype(p), k, k)) p'*X]) *
+        [exp(-p'X); zeros(eltype(p), k, k)],
+    )
+end
+
+@doc raw"""
+    get_basis(M::Stiefel{n,k,ℝ}, p, B::DefaultOrthonormalBasis) where {n,k}
+
+Create the default basis using the parametrization for any $X ∈ T_p\mathcal M$.
+Set $p_\bot \in ℝ^{n\times(n-k)}$ the matrix such that the $n\times n$ matrix of the common
+columns $[p\ p_\bot]$ is an ONB.
+For any skew symmetric matrix $a ∈ ℝ^{k\times k}$ and any $b ∈ ℝ^{(n-k)\times k}$ the matrix
+
+````math
+X = pa + p_\bot b ∈ T_p\mathcal M
+````
+
+and we can use the $\frac{1}{2}k(k-1) + (n-k)k = nk-\frac{1}{2}k(k+1)$ entries
+of $a$ and $b$ to specify a basis for the tangent space.
+using unit vectors for constructing both
+the upper matrix of $a$ to build a skew symmetric matrix and the matrix b, the default
+basis is constructed.
+
+Since $[p\ p_\bot]$ is an automorphism on $ℝ^{n\times p}$ the elements of $a$ and $b$ are
+orthonormal coordinates for the tangent space. To be precise exactly one element in the upper
+trangular entries of $a$ is set to $1$ its symmetric entry to $-1$ and we normalize with
+the factor $\frac{1}{\sqrt{2}}$ and for $b$ one can just use unit vectors reshaped to a matrix
+to obtain orthonormal set of parameters.
+"""
+function get_basis(M::Stiefel{n,k,ℝ}, p, B::DefaultOrthonormalBasis{ℝ}) where {n,k}
+    V = get_vectors(M, p, B)
+    return CachedBasis(B, V)
+end
+
+function get_coordinates!(
+    M::Stiefel{n,k,ℝ},
+    c,
+    p,
+    X,
+    B::DefaultOrthonormalBasis{ℝ},
+) where {n,k}
+    V = get_vectors(M, p, B)
+    c .= [inner(M, p, v, X) for v in V]
```

Comment:
Avoids an allocation
```suggestion
    c .= inner.(Ref(M), Ref(p), V, Ref(X))
```

Reply:
Ah, cool that you also looked at the older code (I just reorganised per metric).